File: db4o.net/Db4objects.Db4o.Tests/Db4objects.Db4o.Tests/Common/Internal/ClassMetadataTypeHandlerIntegrationTestCase.cs
/* Copyright (C) 2004 - 2011  Versant Inc.  http://www.db4o.com */

using System;
using Db4oUnit;
using Db4oUnit.Extensions;
using Db4objects.Db4o.Config;
using Db4objects.Db4o.Foundation;
using Db4objects.Db4o.Internal;
using Db4objects.Db4o.Internal.Delete;
using Db4objects.Db4o.Marshall;
using Db4objects.Db4o.Tests.Common.Internal;
using Db4objects.Db4o.Typehandlers;

namespace Db4objects.Db4o.Tests.Common.Internal
{
	public class ClassMetadataTypeHandlerIntegrationTestCase : AbstractDb4oTestCase
	{
		public class Item
		{
		}

		public class MyReferenceType
		{
		}

		public class MyReferenceTypeHandler : IReferenceTypeHandler
		{
			public virtual void Activate(IReferenceActivationContext context)
			{
			}

			public virtual void Defragment(IDefragmentContext context)
			{
			}

			/// <exception cref="Db4objects.Db4o.Ext.Db4oIOException"></exception>
			public virtual void Delete(IDeleteContext context)
			{
			}

			public virtual void Write(IWriteContext context, object obj)
			{
			}
		}

		/// <exception cref="System.Exception"></exception>
		protected override void Configure(IConfiguration config)
		{
			config.RegisterTypeHandler(new SingleClassTypeHandlerPredicate(typeof(ClassMetadataTypeHandlerIntegrationTestCase.MyReferenceType
				)), new ClassMetadataTypeHandlerIntegrationTestCase.MyReferenceTypeHandler());
		}

		/// <exception cref="System.Exception"></exception>
		protected override void Store()
		{
			Store(new ClassMetadataTypeHandlerIntegrationTestCase.Item());
			Store(new ClassMetadataTypeHandlerIntegrationTestCase.MyReferenceType());
		}

		public virtual void TestIsValueType()
		{
			Pair[] typeDescriptorArray = TypeDescriptors();
			for (int typeDescriptorIndex = 0; typeDescriptorIndex < typeDescriptorArray.Length
				; ++typeDescriptorIndex)
			{
				Pair typeDescriptor = typeDescriptorArray[typeDescriptorIndex];
				ClassMetadata classMetadata = Container().ClassMetadataForObject(typeDescriptor.first
					);
				Assert.AreEqual(((bool)typeDescriptor.second), classMetadata.IsValueType(), classMetadata
					.ToString());
			}
		}

		private Pair[] TypeDescriptors()
		{
			return new Pair[] { Pair(1, true), Pair(new DateTime(), true), Pair("astring", true
				), Pair(new ClassMetadataTypeHandlerIntegrationTestCase.Item(), false), Pair((new 
				int[] { 1 }), false), Pair((new DateTime[] { new DateTime() }), false), Pair((new 
				ClassMetadataTypeHandlerIntegrationTestCase.Item[] { new ClassMetadataTypeHandlerIntegrationTestCase.Item
				() }), false), Pair(new ClassMetadataTypeHandlerIntegrationTestCase.MyReferenceType
				(), false) };
		}

		private Pair Pair(object first, object second)
		{
			return new Pair(first, second);
		}
	}
}

File: db4oj.tests/src/com/db4o/db4ounit/common/internal/ClassMetadataTypeHandlerIntegrationTestCase.java
/* Copyright (C) 2008  Versant Inc.  http://www.db4o.com */

package com.db4o.db4ounit.common.internal;

import java.util.*;

import com.db4o.config.*;
import com.db4o.ext.*;
import com.db4o.foundation.*;
import com.db4o.internal.*;
import com.db4o.internal.delete.*;
import com.db4o.marshall.*;
import com.db4o.typehandlers.*;

import db4ounit.*;
import db4ounit.extensions.*;


public class ClassMetadataTypeHandlerIntegrationTestCase extends AbstractDb4oTestCase {
    
    public static class Item {
    }
    
    public static class MyReferenceType {
    }

    public static class MyReferenceTypeHandler implements ReferenceTypeHandler {

		public void activate(ReferenceActivationContext context) {
		}

		public void defragment(DefragmentContext context) {
		}

		public void delete(DeleteContext context) throws Db4oIOException {
		}

		public void write(WriteContext context, Object obj) {
		}
    }
    
    protected void configure(Configuration config) throws Exception {
        config.registerTypeHandler(
            new SingleClassTypeHandlerPredicate(MyReferenceType.class), 
            new MyReferenceTypeHandler());
    }
    
    protected void store() throws Exception {
        store(new Item());
        store(new MyReferenceType());
    }
    
    public void testIsValueType(){
        for (Pair<Object, Boolean> typeDescriptor : typeDescriptors()) { 
            ClassMetadata classMetadata = container().classMetadataForObject(typeDescriptor.first);
            Assert.areEqual(typeDescriptor.second.booleanValue(), classMetadata.isValueType(), classMetadata.toString());
        }
    }

	private Pair<Object, Boolean>[] typeDescriptors() {
		 return new Pair[] {
			 pair(new Integer(1), true),
			 pair(new Date(), true),
			 pair("astring", true),
			 pair(new Item(), false),
			 pair((new int[] {1}), false),
			 pair((new Date[] {new Date()}), false),
			 pair((new Item[] {new Item()}), false),
			 pair(new MyReferenceType(), false),
		 };
	}
	
	private <TFirst, TSecond> Pair<TFirst, TSecond> pair(TFirst first, TSecond second) {
		return new Pair<TFirst, TSecond>(first, second);
	}

}
